ns3_adapter/src/ns3_client.cpp

```diff
@@ -13,86 +14,134 @@ NS3Client::~NS3Client() {
+    }catch(std::exception e)
+    {
+        ignored_ec = boost::asio::error::invalid_argument;
+        throw e;
     }
-    // else ROS_DEBUG_STREAM( "Connection failed" );
-    else std::cerr << "Connection failed" << std::endl;
 
+    ignored_ec.clear();
 
-    // close();
-    
-    return success;
-}
-
+    io_.reset(new boost::asio::io_service());
+    output_strand_.reset(new boost::asio::io_service::strand(*io_));
+    udp_out_registration_socket_.reset(new boost::asio::ip::udp::socket(*io_,remote_udp_ep_registration_.protocol()));
+    udp_out_broadcasting_socket_.reset(new boost::asio::ip::udp::socket(*io_,remote_udp_ep_broadcasting_.protocol()));
+    work_.reset(new boost::asio::io_service::work(*io_));
 
-bool NS3Client::connect(const std::string &remote_address, unsigned short remote_port,
-                            unsigned short local_port) {
-    boost::system::error_code ignored_ec;
-    return connect(remote_address, remote_port, local_port, ignored_ec);
+    io_thread_.reset(new std::thread([this]()
+                                     {
+                                         boost::system::error_code err;
+                                         io_->run(err);
+                                         if(err)
+                                         {
+                                             onError(err);
+                                         }
+                                     }));
+    running_ = true;
+    onConnect();
+    return true;
 }
 
-bool NS3Client::connect(const std::string &remote_address,
-                                unsigned short remote_port,
-                                 unsigned short local_port,
+
+bool NS3Client::connect_registration_and_broadcasting(const std::string &remote_address,
+                                unsigned short remote_broadcasting_port,
+                                unsigned short remote_registration_port,
+                                 unsigned short local_v2x_port,
+                                 unsigned short local_time_port,
                                  boost::system::error_code &ec)
 {
     //If we are already connected return false
     if(running_) return false;
     //Get remote endpoint
     try
     {
-        remote_udp_ep_ = boost::asio::ip::udp::endpoint(boost::asio::ip::address::from_string(remote_address),remote_port);
+        remote_udp_ep_broadcasting_ = boost::asio::ip::udp::endpoint(boost::asio::ip::address::from_string(remote_address),remote_broadcasting_port);
     }catch(std::exception e)
     {
         ec = boost::asio::error::invalid_argument;
         throw e;
     }
 
-    ec.clear(); 
+    try
+    {
+        remote_udp_ep_registration_ = boost::asio::ip::udp::endpoint(boost::asio::ip::address::from_string(remote_address),remote_registration_port);
+    }catch(std::exception e)
+    {
+        ec = boost::asio::error::invalid_argument;
```

Comment:
Same comment about using const & and about why are we catching and exception just to rethrow the same exception and why are we setting ec just to clear it?

Reply:
I modified it to return false and get out of the function. However, I actually don't think the boost error code is adding much value, so I removed it.